pkg/storage/redis/redis.go

```diff
@@ -208,6 +212,78 @@ func (db *DB) ClearDeleted(_ context.Context, cutoff time.Time) {
 	}
 }
 
+func doNotify(ctx context.Context, psc *redis.PubSubConn, ch chan struct{}) error {
+	switch v := psc.ReceiveWithTimeout(time.Second).(type) {
+	case redis.Message:
+		log.Debug().Str("action", string(v.Data)).Msg("Got redis message")
+		if string(v.Data) != watchAction {
+			return nil
+		}
+		select {
+		case <-ctx.Done():
+			log.Warn().Err(ctx.Err()).Msg("unable to notify channel")
+			return ctx.Err()
+		case ch <- struct{}{}:
+		}
+	case error:
+		log.Debug().Err(v).Msg("redis subscribe error")
+		return v
+	}
+	return nil
+}
+
+func (db *DB) watchLoop(ctx context.Context, ch chan struct{}, eb *backoff.ExponentialBackOff) {
+top:
+	psConn := db.pool.Get()
+	defer psConn.Close()
+	psc := redis.PubSubConn{Conn: psConn}
+	if err := psc.PSubscribe("__keyspace*__:" + db.versionSet); err != nil {
+		log.Error().Err(err).Msg("failed to subscribe to version set channel")
+		return
+	}
+
+	for {
+		err, ok := doNotify(ctx, &psc, ch).(net.Error)
+		if !ok && err != nil {
+			log.Error().Err(ctx.Err()).Msg("failed to notify channel")
+			return
+		}
+		if ok && err.Timeout() {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(eb.NextBackOff()):
+			}
+			goto top
+		}
+	}
+}
+
+// Watch returns a channel to the caller, when there is a change to the version set,
+// sending message to the channel to notify the caller.
+func (db *DB) Watch(ctx context.Context) chan struct{} {
+	ch := make(chan struct{})
+	go func() {
```

Comment:
Are you concerned about starting this go-routine without knowing when it will stop?

Reply:
Why "without knowing"? The `ctx` is passed to make sure that the caller can control when to signal the goroutine to stop.

Reply:
Got it. I had misread the code.